// react-native-paper/lib/commonjs/core/theming.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.withTheme = exports.useTheme = exports.getTheme = exports.defaultThemesByVersion = exports.ThemeProvider = exports.DefaultTheme = void 0;

var _reactThemeProvider = require("@callstack/react-theme-provider");

var _themes = require("../styles/themes");

const DefaultTheme = _themes.MD3LightTheme;
exports.DefaultTheme = DefaultTheme;
const {
  ThemeProvider,
  withTheme,
  useTheme
} = (0, _reactThemeProvider.createTheming)(DefaultTheme);
exports.useTheme = useTheme;
exports.withTheme = withTheme;
exports.ThemeProvider = ThemeProvider;
const defaultThemesByVersion = {
  2: {
    light: _themes.MD2LightTheme,
    dark: _themes.MD2DarkTheme
  },
  3: {
    light: _themes.MD3LightTheme,
    dark: _themes.MD3DarkTheme
  }
};
exports.defaultThemesByVersion = defaultThemesByVersion;

const getTheme = function () {
  let isDark = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : false;
  let isV3 = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : true;
  const themeVersion = isV3 ? 3 : 2;
  const scheme = isDark ? 'dark' : 'light';
  return defaultThemesByVersion[themeVersion][scheme];
};

exports.getTheme = getTheme;
