// react-native-paper/lib/module/components/Typography/AnimatedText.js
function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

import * as React from 'react';
import { Animated, I18nManager, StyleSheet, Platform } from 'react-native';
import { withTheme } from '../../core/theming';
import { MD3TypescaleKey } from '../../types';

/**
 * Animated text component which follows styles from the theme.
 *
 * @extends Text props https://reactnative.dev/docs/text#props
 */
function AnimatedText(_ref) {
  let {
    style,
    theme,
    variant,
    ...rest
  } = _ref;
  const writingDirection = I18nManager.isRTL ? 'rtl' : 'ltr';

  if (theme.isV3 && variant) {
    const stylesByVariant = Object.keys(MD3TypescaleKey).reduce((acc, key) => {
      const {
        fontSize,
        fontWeight,
        lineHeight,
        letterSpacing,
        fontFamily
      } = theme.typescale[key];
      return { ...acc,
        [key]: { ...(Platform.OS === 'android' && {
            fontFamily
          }),
          fontSize,
          fontWeight,
          lineHeight: lineHeight,
          letterSpacing,
          color: theme.colors.onSurface
        }
      };
    }, {});
    const styleForVariant = stylesByVariant[variant];
    return /*#__PURE__*/React.createElement(Animated.Text, _extends({}, rest, {
      style: [styleForVariant, styles.text, {
        writingDirection
      }, style]
    }));
  } else {
    return /*#__PURE__*/React.createElement(Animated.Text, _extends({}, rest, {
      style: [styles.text, { ...(!theme.isV3 && theme.fonts.regular),
        color: theme.isV3 ? theme.colors.onSurface : theme.colors.text,
        writingDirection
      }, style]
    }));
  }
}

const styles = StyleSheet.create({
  text: {
    textAlign: 'left'
  }
});
export default withTheme(AnimatedText);
